### PR TITLE
Activate/Expire Holds & Notify Users

### LINKED ISSUE #4090: Add ChargeManager#activate_holds
## Implementation Notes

Do nothing if there's no available charge slots.

For every available charge slot, find the next inactive hold and give it an expiration time 1 hour in the future. Send an email to the user notifying them that they have the item on hold. The text of the email should be something like

Subject: `Available for Digital Checkout: < resource title >`

Body:
> < resource title > is now available for check out for the next 1 hour. In that time please visit the catalog here(link to catalog) to view it.

We can probably do better than that, but that's what I got at the moment.

ALSO, if there are any expired holds delete them from the hold_queue and notify the appropriate user that their hold's expired.

### PR DESCRIPTION
**Pull request by tpendragon**
Closes #4090 

This also includes some refactors to move around responsibilities a bit.

The emails aren't pretty, but there's another ticket to look at that: #4117

### CODE REVIEW

**Comment by tpendragon** [comment]
The email is VERY basic right now:

Screen Shot 2020-07-17 at 1 50 44 PM

File: app/resources/cdl/charge_manager.rb [hunk 1]
```diff
@@ -84,8 +58,47 @@ def create_hold(netid:)
       change_set_persister.save(change_set: change_set)
     end
 
-    def available_item_id
-      (item_ids - resource_charge_list.charged_items.map(&:item_id)).first
+    def activate_holds!
+      @resource_charge_list = expire_holds!
+      return unless available_charge_slot? && resource_charge_list.pending_or_active_holds.present?
```
Comment (hackmastera) [inline app/resources/cdl/charge_manager.rb:63]: why is it `resource_charge_list` here but `@resource_charge_list` above?

File: app/resources/cdl/charge_manager.rb [hunk 2]
```diff
@@ -84,8 +58,47 @@ def create_hold(netid:)
...
+      end
+      change_set = CDL::ResourceChargeListChangeSet.new(resource_charge_list)
+      change_set.validate(hold_queue: resource_charge_list.hold_queue)
+      change_set_persister.save(change_set: change_set)
+      activated_holds.each do |hold|
+        notify_hold_active(hold: hold)
+      end
+    end
+
+    def activate_hold_count
```
Comment (hackmastera) [inline app/resources/cdl/charge_manager.rb:78]: Is there a better name for this? It seems like something is going to get activated when it's called.

Comment (hackmastera) [reply app/resources/cdl/charge_manager.rb:78]: It seems to be a count of items available to hold. Not sure how to phrase this more clearly.

File: app/resources/cdl/charge_manager.rb [hunk 3]
```diff
@@ -84,8 +58,47 @@ def create_hold(netid:)
...
 
-    def available_item_id
-      (item_ids - resource_charge_list.charged_items.map(&:item_id)).first
+    def activate_holds!
+      @resource_charge_list = expire_holds!
+      return unless available_charge_slot? && resource_charge_list.pending_or_active_holds.present?
+      return unless activate_hold_count.positive?
+      activated_holds = Array.new(activate_hold_count).map do
+        hold = resource_charge_list.pending_holds.first
+        hold.expiration_time = 1.hour.from_now
```
Comment (hackmastera) [inline app/resources/cdl/charge_manager.rb:67]: where is this saved?

File: app/resources/cdl/charge_manager.rb [hunk 4]
```diff
@@ -84,8 +58,47 @@ def create_hold(netid:)
       change_set_persister.save(change_set: change_set)
     end
 
-    def available_item_id
-      (item_ids - resource_charge_list.charged_items.map(&:item_id)).first
+    def activate_holds!
+      @resource_charge_list = expire_holds!
+      return unless available_charge_slot? && resource_charge_list.pending_or_active_holds.present?
+      return unless activate_hold_count.positive?
+      activated_holds = Array.new(activate_hold_count).map do
```
Comment (hackmastera) [inline app/resources/cdl/charge_manager.rb:65]: 👍

File: app/resources/cdl/charge_manager.rb (hunk 3 above)
Comment (tpendragon) [reply app/resources/cdl/charge_manager.rb:67]: The holds are mutated in place and then saved below.

**Comment by tpendragon** [comment]
@hackmastera Should be ready for re-review.

File: app/resources/cdl/charge_manager.rb (hunk 3 above)
Comment (hackmastera) [reply app/resources/cdl/charge_manager.rb:67]: okay they're held by reference in the hold_queue array.

**Pull request merged**